File: binary_search/intervals_counter/intervals_counter.cpp
#include <vector>
using namespace std;

class IntervalsCounter
{ // LeetCode Q.2276.
private:
    vector<pair<int, int>> ranges; // Format: {inclusive left, inclusive right}.
    int covered_nums = 0;

public:
    IntervalsCounter() {}

    void add(int left, int right)
    {
        int idx = upper_bound(
                      ranges.begin(), ranges.end(), left,
                      [](int val, const pair<int, int> &p)
                      { return val < p.first; }) -
                  ranges.begin();

        if (idx > 0 && ranges[idx - 1].second + 1 >= left)
        {             // Merge prev range.
            idx -= 1; // Insertion idx changes.
            covered_nums -= ranges[idx].second + 1 - ranges[idx].first;
            right = max(right, ranges[idx].second);
            ranges[idx].second = right;
            left = ranges[idx].first;
        }
        else
        {
            ranges.insert(ranges.begin() + idx, {left, right});
        }

        while (idx + 1 < ranges.size() && ranges[idx + 1].first - 1 <= right)
        {
            covered_nums -= ranges[idx + 1].second + 1 - ranges[idx + 1].first;
            right = max(right, ranges[idx + 1].second);
            ranges[idx].second = right;
            ranges.erase(ranges.begin() + idx + 1); // Next range is merged.
        }

        covered_nums += right + 1 - left;
    }

    int count()
    {
        return covered_nums;
    }
};

File: binary_search/intervals_counter/intervals_counter.py
from bisect import bisect_right


class IntervalsCounter:  # LeetCode Q.2276.
    def __init__(self):
        self.ranges = []  # Format: [inclusive left, inclusive right].
        self.covered_nums = 0

    def add(self, left: int, right: int) -> None:
        idx = bisect_right(self.ranges, [left, right])

        if idx > 0 and self.ranges[idx - 1][1] + 1 >= left:  # Merge prev range.
            idx -= 1  # Insertion idx changes.
            self.covered_nums -= self.ranges[idx][1] + 1 - self.ranges[idx][0]
            right = max(right, self.ranges[idx][1])
            self.ranges[idx][1] = right
            left = self.ranges[idx][0]

        else:
            self.ranges.insert(idx, [left, right])

        while idx + 1 < len(self.ranges) and self.ranges[idx + 1][0] - 1 <= right:
            self.covered_nums -= self.ranges[idx + 1][1] + 1 - self.ranges[idx + 1][0]
            right = max(right, self.ranges[idx + 1][1])
            self.ranges[idx][1] = right
            self.ranges.pop(idx + 1)  # Next range is merged.

        self.covered_nums += right + 1 - left

    def count(self) -> int:
        return self.covered_nums
